### ISSUE TITLE
Polkanomicon

### ISSUE THREAD

**Issue by pepyakin** [paritytech/polkadot#6143]
The original intent of the implementer's guide is to describe Polkadot in enough detail to be implemented by other teams. 

However, from my perception, this document is not really used this way. This is not to say that the document is useless. On the contrary, it was of immense value for onboarding and pointing to different protocol elements. While it is still useful for other implementers, their staple should be the polkadot spec or similar documents.

I think it might be time for us to admit that and embrace the actual purpose. Therefore,

I want to propose converting the implementer's guide to a dev guide or simply the book.

By that, I don't only mean changing the name but embracing the pattern widely used in the open-source ecosystem. Below are distinct examples of books:

1. https://rustc-dev-guide.rust-lang.org/
1. https://docs.wasmtime.dev/
1. https://book.getfoundry.sh/

Applied to our case, I think the book should serve people routinely contributing to the Polkadot (core) project. I think good succint description should be: this is the book that existing or prospective [Polkadot Fellows](https://forum.polkadot.network/t/calling-polkadot-core-developers/506) might find on their tables. The contents we have right now are not very far for the desired place. 

Several things:

1. Description of the architecture. This is already mostly in place. I think we need to change the level of description though. Some of the modules are described in such detail that it will be trivial to translate those into code. This is prone to outdate and unnecessary burdens. The level of detail should be so that after reading it, the corresponding piece of code will be trivial to read. On the other hand, I feel there is more to say about the high-level architecture.
1. How to work with code. Coding conventions. Requirements for documentation (i.e. update the guide if the code changes). General tenets we strive when writing code. Bonus points, if it's formulated in a way that is easy to point out in PRs. 
1. How to work with the protocol. Principles working with and designing the protocol, like ultimate requirements for determinism, versioning, what to do if a primitive type is required. 
1. Description of the userland. That includes PDK (parachain development kit) information. Cumulus section: brief description, description of the common connection points (e.g. how DMP messages are getting from relay-chain to parachain), etc. "Userland docs": what interface polkadot requires/provides from/to PVFs. Docs required to write other PDKs.
1. Working with the repo. How to create PR. What labels are required, and criterias for how to use labels ("is this B-silent"). The bot commands. Troubleshooting. What to do if the CI or the bot does not work.
1. Building, Running, Testing. How to effectively run the node for testing and monitoring. Tips and tricks. What to look for. Common troubleshooting. Also bonus points if it's forumulated in a way that could be pointed to from the issues.
1. Branch out for links for additional information, other developer-related useful links, etc. For example, polkadot specification, wasm specification, more info about VRF.

Here I only propose the vision. Obviously, it will take some time to get there, but getting on the same page about the vision could be the first step.

**Comment by eskimor** [comment]
Sounds all pretty reasonable, mostly what I have in mind as well. Apart from general architecture and overall design, what I really want to see in the guide is the why's and the reasoning behind the code. E.g. threat models and how the code tackles them.

**Comment by tdimitrov** [comment]
I generally agree but I believe points 2, 5 and partially 6 should be in a separate guide/document. **The book** should be about Polkadot itself - protocols and archtecture.
Code style, branching, etc better be in a shorter contributors.md in the repo.

**Comment by pepyakin** [comment]
> Sounds all pretty reasonable, mostly what I have in mind as well. Apart from general architecture and overall design, what I really want to see in the guide is the why's and the reasoning behind the code. E.g. threat models and how the code tackles them.

@eskimor yes, agreed, had same in mind! 

> Code style, branching, etc better be in a shorter contributors.md in the repo.

@tdimitrov I see where you are coming from, and at other times I would agree with you. Not now, though.

The reason why I think it's better is that having those materials in the book will increase their discoverability. People who went for other reasons there more likely will discover such a document in contrast to disparate documents scattered around the repo. I also don't see a downside: the "cost" of those docs being in the book is an entry in the summary.

Having more discoverability the information will disseminate better. Also, the quality should improve since it will receive more eyes. Conversion from people who spotted a mistake (or any other contribution suggestion) to make the PRs will be better.

**Comment by rphmeier** [comment]
I agree that the original approach to the implementers' guide had too much detail. It was simultaneously being used as both a design document and reference, which was convenient  before code was written and inconvenient once code was done being implemented. A first pass would distill the pages down to key concepts and eliminate references to specific algorithms or implementation details.

**Comment by bkchr** [comment]
Overall liking the idea!

> * How to work with the protocol. Principles working with and designing the protocol, like ultimate requirements for determinism, versioning, and what to do if a primitive type is required.
> 
> * Description of the userland. That includes PDK (parachain development kit) information. Cumulus section: brief description, description of the common connection points (e.g. how DMP messages are getting from relay-chain to parachain), etc. "Userland docs": what interface polkadot requires/provides from/to PVFs. Docs required to write other PDKs.

IMO, these things should go into the spec. 

We should clearly try to prevent to have too many duplicate information in the spec and this guide here. Duplicate information just leads to stuff getting outdated at one point. Most of the high level information about how approval voting, etc works should go into the spec. Then followed by the "pure" spec of the protocol. Architecture etc should stay here in the Polkadot repo, because that isn't important in the spec.
